src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/VariantAnnotatorEngine.java

```diff
@@ -0,0 +1,296 @@
+            if (addAnnot.test(annotationType)){
+                final Map<String, Object> annotationsFromCurrentType = annotationType.annotate(ref, newGenotypeAnnotatedVC, stratifiedPerReadAlleleLikelihoodMap);
+                if ( annotationsFromCurrentType != null ) {
+                    infoAnnotMap.putAll(annotationsFromCurrentType);
+                }
+            }
+        }
+
+        // create a new VC in the with info and genotype annotations
+        final VariantContext annotated = builder.attributes(infoAnnotMap).make();
+
+        // annotate db occurrences
+        return variantOverlapAnnotator.annotateOverlaps(features, variantOverlapAnnotator.annotateRsID(features, annotated));
+    }
+
+    /**
+     * @param stratifiedPerReadAlleleLikelihoodMap par-sample map of likelihoods per read. May be null
+     */
+    private GenotypesContext annotateGenotypes(final ReferenceContext ref,
+                                               final VariantContext vc,
+                                               final Map<String, PerReadAlleleLikelihoodMap> stratifiedPerReadAlleleLikelihoodMap,
+                                               final Predicate<VariantAnnotation> addAnnot) {
+        if ( genotypeAnnotations.isEmpty() ) {
+            return vc.getGenotypes();
+        }
+
+        final GenotypesContext genotypes = GenotypesContext.create(vc.getNSamples());
+        for ( final Genotype genotype : vc.getGenotypes() ) {
+            PerReadAlleleLikelihoodMap perReadAlleleLikelihoodMap = null;
+            if (stratifiedPerReadAlleleLikelihoodMap != null) {
+                perReadAlleleLikelihoodMap = stratifiedPerReadAlleleLikelihoodMap.get(genotype.getSampleName());
+            }
+
+            final GenotypeBuilder gb = new GenotypeBuilder(genotype);
+            for ( final GenotypeAnnotation annotation : genotypeAnnotations) {
+                if (addAnnot.test(annotation)) {
+                    annotation.annotate(ref, vc, genotype, gb, perReadAlleleLikelihoodMap);
+                }
+            }
+            genotypes.add(gb.make());
+        }
+
+        return genotypes;
+    }
+
+    private static final class AnnotationManager {
+
+        private final List<String> annotationGroupsToUse;
+        private final List<String> annotationsToUse;
+        private final List<String> annotationsToExclude;
+
+        private AnnotationManager(final List<String> annotationGroupsToUse, final List<String> annotationsToUse, final List<String> annotationsToExclude){
+            this.annotationGroupsToUse = annotationGroupsToUse;
+            this.annotationsToUse = annotationsToUse;
+            this.annotationsToExclude = annotationsToExclude;
+
+            final Set<String> allAnnotationNames = new HashSet<>(AnnotationManager.getAllAnnotationNames());
+            final Set<String> unknownAnnots = Sets.difference(new HashSet<>(annotationsToUse), allAnnotationNames);
+            if (!unknownAnnots.isEmpty()){
```

Comment:
Please extract an `annotationExists()` helper method, and use it here and below.